### ISSUE TITLE
feature request: equivalent of gdb display command

### ISSUE THREAD

**Issue by sam-github** [go-delve/delve#1256]
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)? 1.0.0
2. What version of Go are you using? (`go version`)? 1.9.2
3. What operating system and processor architecture are you using? ox x, i386

I would like to see an equivalent to the gdb display command, see ftp://ftp.gnu.org/old-gnu/Manuals/gdb/html_chapter/gdb_9.html#SEC57

Basically, when stepping through code and trying to understand how some set of variables (or an expression) is changing its values as the debugger executes lines, I find it very common to after every line print a set of variables to see what the current values are. This takes a couple up arrow and RETURN key sequences, but its pretty handy to use the display command to say "print all these every time the debugger stops" and then I can just press a single key to step over line after line, watching the variables to see how their values mutate.

With gdb, I use display more than print, it seems to almost always be what I want to do.

Thanks for dlv! Cheers,Sam

**Comment by aarzilli** [comment]
This is a reasonable feature request. If you want to implement this yourself the entire implementation of the command line client rests in `pkg/terminal/command.go`. In particular you are interested in two functions `(*Commands).cont` and `continueUntilCompleteNext`, if you keep a list of expressions somewhere then print all of them (see `printVar`) before those two functions exit you will have the `display` command.

**Comment by sam-github** [comment]
I'll give it a shot and see how far I get, thanks.

**Comment by Omie** [comment]
any progress or is this up for grab @sam-github ? I was hoping to find something to work on here!

**Comment by sam-github** [comment]
Its up for grab, sorry, I'm too busy on other  things.

**Comment by Omie** [comment]
@sam-github  thanks for getting back. I will look into it.

@aarzilli @derekparker long time! I hope you folks are doing great.

Would like to get few things answered.

To begin with, I suppose I can use linked document as requirement as is. Will copy the text as part of documentation, PR description etc (with reference to original doc). Good to go?

how to handle multi word commands? gdb's doc has something like `enable display ... `, `disable display ...`. I didn't see any multi-word commands for delve, how should it be handled here? `display-enable` `display-disable` would do? Or anything different/better?

I see that there's a way to save delve's running session config. Should this be part of it?

Will have more questions as I get involved, just catching up on code :)

**Comment by aarzilli** [comment]
> To begin with, I suppose I can use linked document as requirement as is. Will copy the text as part of documentation, PR description etc (with reference to original doc). Good to go?

Well we don't have the `/fmt` syntax anywhere and also we don't do multiword commands, so no.

>how to handle multi word commands?

I'd make them flags of display. Have `display <expr>` and `display -- <expr>` add an expression and then `display -d <num>` to delete a display expression, &c. I don't know if @derekparker has different ideas.

> I see that there's a way to save delve's running session config. Should this be part of it?

We don't save any per-target stuff (like breakpoints for example), I'd say no, at least for now.

**Comment by Omie** [comment]
> I'd make them flags of display. Have `display <expr>` and `display -- <expr>` add an expression and then `display -d <num>` to delete a display expression, &c. I don't know if @derekparker has different ideas.

this sounds good

> We don't save any per-target stuff (like breakpoints for example), I'd say no, at least for now.

cool, thanks

**Comment by Omie** [comment]
As suggested above, we would like to have equivalent of `display` command from GDB. Working of display command from GDB is explained at: [ftp://ftp.gnu.org/old-gnu/Manuals/gdb/html_chapter/gdb_9.html#SEC57](ftp://ftp.gnu.org/old-gnu/Manuals/gdb/html_chapter/gdb_9.html#SEC57), copying the text here for reference

```
Automatic display

If you find that you want to print the value of an expression frequently (to see how it changes), you might want to add it to the automatic display list so that GDB prints its value each time your program stops. Each expression added to the list is given a number to identify it; to remove an expression from the list, you specify that number. The automatic display looks like this:

2: foo = 38
3: bar[5] = (struct hack *) 0x3804
This display shows item numbers, expressions and their current values. As with displays you request manually using x or print, you can specify the output format you prefer; in fact, display decides whether to use print or x depending on how elaborate your format specification is--it uses x if you specify a unit size, or one of the two formats (`i' and `s') that are only supported by x; otherwise it uses print.

display expr
Add the expression expr to the list of expressions to display each time your program stops. See section Expressions. display does not repeat if you press RET again after using it.
display/fmt expr
For fmt specifying only a display format and not a size or count, add the expression expr to the auto-display list but arrange to display it each time in the specified format fmt. See section Output formats.
display/fmt addr
For fmt `i' or `s', or including a unit-size or a number of units, add the expression addr as a memory address to be examined each time your program stops. Examining means in effect doing `x/fmt addr'. See section Examining memory.
For example, `display/i $pc' can be helpful, to see the machine instruction about to be executed each time execution stops (`$pc' is a common name for the program counter; see section Registers).

undisplay dnums...
delete display dnums...
Remove item numbers dnums from the list of expressions to display. undisplay does not repeat if you press RET after using it. (Otherwise you would just get the error `No display number ...'.)
disable display dnums...
Disable the display of item numbers dnums. A disabled display item is not printed automatically, but is not forgotten. It may be enabled again later.
enable display dnums...
Enable display of item numbers dnums. It becomes effective once again in auto display of its expression, until you specify otherwise.
display
Display the current values of the expressions on the list, just as is done when your program stops.
info display
Print the list of expressions previously set up to display automatically, each one with its item number, but without showing the values. This includes disabled expressions, which are marked as such. It also includes expressions which would not be displayed right now because they refer to automatic variables not currently available.
If a display expression refers to local variables, then it does not make sense outside the lexical context for which it was set up. Such an expression is disabled when execution enters a context where one of its variables is not defined. For example, if you give the command display last_char while inside a function with an argument last_char, GDB displays this argument while your program continues to stop inside that function. When it stops elsewhere--where there is no variable last_char---the display is disabled automatically. The next time your program stops where last_char is meaningful, you can enable the display expression once again.

```

As learned from the gdb's display command, this single command is responsible for multiple operations, viz,
- add a new expression for automatic display
- remove a selected expression from automatic display
- evaluate relevant expressions every time debugger stops at prompt
- disable a selected expression without removing it
- enable a selected expression
- automatically enable/disable expression based on scope
- display current set of expressions (without evaluating them)
- run all the expressions again on manual invocation, output would be same as automatic display

To incorporate this in delve, we have to modify it a little.
- delve does not have /fmt syntax, so that will not be incorporated here
- delve does not have multiple word commands, we'd rather use `display` command with flags

Considering all of above, here's how we could do it

| Command | Description |
| --- | --- |
| display | evaluate current set of expressions and print results |
| display -a `<expr>` | Add given expression to automatic display |
| display -r `<expr-id>` | Remove expression from automatic display |
| display -d `<expr-id>` | Disable selected expression |
| display -e `<expr-id>` | Enable selected expression |
| display -l | Show current set of expressions without evaluating them |

I have mixed thoughts about adding `display -x <expr-id>` which can evaluate a single expression.
And also, what should be the default behavior of the command? 

Output of `display` could look like below. Note that, this is only about what information to be printed and not the exact formatting. Formatting depends on pretty printing library that is used. We'd make it render similar like list of breakpoints is displayed.

| Id | Expression | Value |
| --- | --- | --- |
| 1 | foo | 1234 |
| 2 | bar[5] | "bazzinga" |


Output of `display -l` could look like below

| Id | Expression | Enabled | Scope |
| --- | --- | --- | --- |
| 1 | foo | true | `<func-name-or-some-id>` |
| 2 | bar | false | `<func-name-or-some-id>` |


Note: This is my initial thought that if we simply store some sort of unique ID for scope along with the expression, then it'd be pretty straight forward to evaluate expressions that fit current execution scope. This would help in avoiding local variables that now are out of scope. I haven't concluded on what-and-how to do this as I am not yet very much familiar with the code. Currently it only answers **why** :)

Also, I have noticed that there's already a command called `on` that has little similar expected behavior. Wondering if `display` can be merged with `on`. This can certainly be decided later once the functionality for `display` is in place.

any suggestion or feedback is appreciated.

**Comment by sam-github** [comment]
Behaviour described above looks good to me. I'm OK with display default being same as `gdb`'s, or for it to print a summary of the usage - not sure what is more idiomatic for `delve`.

I think `on` is pretty different in use. Generally you set breakpoints, then let the program continue running. `display` is more often used while single stepping through source. But, the displayed expressions should evaluate when a BP is reached.

**Comment by aarzilli** [comment]
I think this is fine.

**Comment by AdamVanScyoc** [comment]
Just curious, why was this never merged and the issue closed?

**Comment by sam-github** [comment]
This is the issue, its still open. There is no PR to merge.

**Comment by Omie** [comment]
Ah, sorry, my bad. I have a functional changeset in my fork, tests are pending.
IIRC what happened that time: I went off to hometown for couple of weeks, got back and fell sick, afterwards got cought up with life + (saturated) work and this starved. I'm heavily occupied pretty much since then.

**Comment by mewmew** [comment]
Now that #1917 has been merged, is this issue ready to be closed? I tried the `display` command for a sample binary and it seems to work well. Are there any known issues or aspects left to implement?

Also, thanks for everyone who's been working on this issue. It's great to have in Delve!

**Comment by aarzilli** [comment]
Yes, it can be closed.
